### PR TITLE
Add ALTER TABLE ... RENAME TO feature

### LINKED ISSUE #95: Add ALTER TABLE ... RENAME TO ...
This is a sub-issue of #94 that focuses on renaming an existing table.

We should be able to do the following:
```
genji> CREATE TABLE foo;
genji> INSERT INTO foo VALUES {name: "John Doe"};
genji> ALTER TABLE foo RENAME TO bar;
genji> SELECT * FROM bar;
{
  "name": "John Doe"
}
genji> SELECT * FROM foo;
table not found
```
If the table `bar` already exists, we should return an error.

### PR DESCRIPTION
**Pull request by yaziine**
This PR fixes #95. 

We can now rename a table and all indexes linked to the table are updated as well.

### CODE REVIEW

File: database/config.go [hunk 1]
```diff
@@ -158,20 +158,23 @@ func (t *tableInfoStore) Insert(tableName string, info *TableInfo) ([]byte, erro
...
 	var id [6]byte
-	for {
-		id = generateStoreID()
-		_, err = t.st.Get(id[:])
-		if err == nil {
-			// A store with this id already exists.
-			// Let's generate a new one.
-			continue
+	// Generate a new storeID if the one in the info is empty.
+	if info.storeID == id {
```
Comment (asdine) [inline database/config.go:162]: To avoid this logic and since it's only useful the first time we create a table, maybe we should generate the store id outside of this function, in the `Insert` method.

File: database/config.go [hunk 2]
```diff
@@ -345,6 +348,17 @@ func (t *indexStore) Get(indexName string) (*IndexConfig, error) {
 	return &idxopts, nil
 }
 
+func (t *indexStore) Update(cfg IndexConfig) error {
```
Comment (asdine) [inline database/config.go:351]: I think `Replace` would be a better name because we are literally replacing the IndexConfig by another one. Also, it's consistent with TableInfoStore#Replace

File: database/table.go [hunk 3]
```diff
@@ -21,6 +21,11 @@ type Table struct {
 	infoStore *tableInfoStore
 }
 
+// Name returns the name of the table.
```
Comment (asdine) [inline database/table.go:24]: I suppose this is a reordering of the file? Is there any other change? If so we can keep it as is, otherwise please do the reordering in a separate PR otherwise it gets extremely difficult to review.

File: database/transaction.go [hunk 4]
```diff
@@ -103,6 +103,39 @@ func (tx Transaction) GetTable(name string) (*Table, error) {
 	}, nil
 }
 
+// RenameTable renames a table named old to new.
+// If it doesn't exist, returns ErrTableNotFound.
+func (tx Transaction) RenameTable(old, new string) error {
```
Comment (asdine) [inline database/transaction.go:108]: We should avoid shadowing keywords, I'm surprised the linter didn't yell at you. Perhaps something like `oldName`, `newName` ?

File: database/transaction.go [hunk 5]
```diff
@@ -103,6 +103,39 @@ func (tx Transaction) GetTable(name string) (*Table, error) {
 	}, nil
 }
 
+// RenameTable renames a table named old to new.
```
Comment (asdine) [inline database/transaction.go:106]: Renaming is always from something to something else.
```suggestion
// RenameTable renames a table.
```

File: database/transaction.go [hunk 6]
```diff
@@ -103,6 +103,39 @@ func (tx Transaction) GetTable(name string) (*Table, error) {
...
+// If it doesn't exist, returns ErrTableNotFound.
```
Comment (asdine) [inline database/transaction.go:107]:
```suggestion
// If it doesn't exist, it returns ErrTableNotFound.
```

File: database/transaction.go [hunk 7]
```diff
@@ -170,6 +203,16 @@ func (tx Transaction) CreateIndex(opts IndexConfig) error {
 	return tx.indexStore.Insert(opts)
 }
 
+// UpdateIndex updates the index cfg referenced by cfg.IndexName.
+func (tx Transaction) UpdateIndex(cfg IndexConfig) error {
```
Comment (asdine) [inline database/transaction.go:207]: I don't think this should be exported, it seems really specific to the renaming feature and can be dangerous in other contexts. There doesn't seem to be a reason to replace an index config other than for renaming its related table.

File: database/table.go (hunk 3 above)
Comment (yaziine) [reply database/table.go:24]: It's only a reordering. I did that in a separate commit for precisely ease the review.

File: database/transaction.go (hunk 4 above)
Comment (yaziine) [reply database/transaction.go:108]: The linter didn't tell me anything regarding this. Does your tell you something?

File: database/transaction.go (hunk 7 above)
Comment (yaziine) [reply database/transaction.go:207]: Yeah very good point!

File: database/transaction.go (hunk 4 above)
Comment (asdine) [reply database/transaction.go]: It doesn't, that's weird. Though it's a bad habit shadowing keywords, even if that sounds good in that context, plus being explicit in a function signature doesn't hurt

File: database/table.go (hunk 3 above)
Comment (asdine) [reply database/table.go:24]: Cool, thanks

File: database/config.go (hunk 1 above)
Comment (yaziine) [reply database/config.go]: Done

File: database/config.go [hunk 8]
```diff
@@ -414,6 +381,17 @@ func (t *indexStore) Get(indexName string) (*IndexConfig, error) {
 	return &idxopts, nil
 }
 
+func (t *indexStore) Replace(cfg IndexConfig) error {
```
Comment (asdine) [inline database/config.go:384]: When replacing something, the API must explicitly state what is the key used to identify the thing we are trying to replace, even if that information is already contained in the value of the second parameter.

```suggestion
func (t *indexStore) Replace(idxName string, cfg IndexConfig) error {
```

File: database/transaction.go [hunk 9]
```diff
@@ -66,23 +69,53 @@ func (tx *Transaction) Promote() error {
...
 	}
-	sid, err := tx.tableInfoStore.Insert(tx.Tx, name, info)
+
+	var found bool = true
+	var id [6]byte
+	for found {
+		id = generateStoreID()
+
+		found = false
+		for _, ti := range tx.tableInfoStore.tableInfos {
```
[the remark this answers is not in the document]
Comment (yaziine) [reply database/transaction.go:83]: Good catch! 👍
